include/net/socket.h

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __NET_SOCKET_H
+#define __NET_SOCKET_H
+
+#include <sys/types.h>
+#include <zephyr/types.h>
+#include <net/net_ip.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+int zsock_socket(int family, int type, int proto);
+int zsock_close(int sock);
+int zsock_bind(int sock, const struct sockaddr *addr, socklen_t addrlen);
+int zsock_connect(int sock, const struct sockaddr *addr, socklen_t addrlen);
+int zsock_listen(int sock, int backlog);
+int zsock_accept(int sock, struct sockaddr *addr, socklen_t *addrlen);
+ssize_t zsock_send(int sock, const void *buf, size_t len, int flags);
+ssize_t zsock_recv(int sock, void *buf, size_t max_len, int flags);
```

Comment:
> It's often a standard module convention to have symbols declared in their own header. Makes it easy to find things. 

Well, that's actually mu biggest concern - if I'd see `socket.h`, I would never think there's also `zsocket.h`, or vice-versa. Zephyr doesn't split IPv4 and IPv6 headers. lwIP defines both native prefixed functions (lwip_recv(), etc.) and POSIXy macros in the same header, sockets.h. And in this case, it's also the single module, it just defines 2 API sets, one being alias of another (but it's the same API).

> These are Zephyr APIs, which may have divergence from true POSIX APIs (declared in socket.h).

Well, there're no "true POSIX APIs", only "BSD Sockets like API" ("like" is important). True POSIX APIs live in <sys/socket.h>, Zephyr's BSD Sockets like API lives in <net/socket.h>.

> It makes it easier to implement a new socket provider for the socket APIs for offloading. zsocket could just be one (the default) socket provider. Socket.h (or socket.c) would be the switch layer.

Ok, this finally explains why you suggest this change! But why do you want to do it this way? I imagined it would be done just the same way as offloading already implemented in net_context - that each individual zsock_*() will be the place of switching, and that offloading will be supported per netif.

You instead seem to propose that socket offloading will be global, not per netif. Rereading https://jira.zephyrproject.org/browse/ZEP-2271?focusedCommentId=19002&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-19002 , that seems to be what's written in the user story, but if you don't mind, I'd like to get an ack form @jukkar and @tbursztyka that it's ok to proceed with such design right away (vs further discuss it), as it leaves many questions open, e.g. why we have 2 inconsistently implemented offloading mechanisms. Thanks.